### PR TITLE
Logical links

### PR DESCRIPTION
**Pull request by FJanssen-TNO**
Logical links are allowed to be used to prevent the creation of many unnecessary (short) pipes and cables and there corresponding variables. 

Almost all tests are running, just a few more need to be sorted out.

### CODE REVIEW

File: src/mesido/esdl/esdl_model_base.py [hunk 1]
```diff
@@ -388,6 +475,19 @@ def _esdl_convert(self, converter: _AssetToComponentBase, assets: Dict, prefix:
...
                     if conn in connections or tuple(reversed(conn)) in connections:
                         continue
-
-                    self.connect(port_map[port.id], port_map[connected_to.id])
+                    if (
+                        asset.asset_type == "Pipe"
+                        or asset.asset_type == "ElectricityCable"
+                        or assets[
+                            name_to_id_map[port_map[connected_to.id].name.split(".")[0]]
+                        ].asset_type
```
Comment (FJanssen-TNO) [inline src/mesido/esdl/esdl_model_base.py:483]: maybe add a comment on the asset type check in the ports, to make it easier understand when logical links should be created or normal connections.

File: src/mesido/pycml/component_library/milp/electricity/electricity_base.py [hunk 2]
```diff
@@ -13,8 +13,8 @@ def __init__(self, name, **modifiers):
         super().__init__(name, **modifiers)
 
         self.add_variable(Variable, "Power")
-        self.add_variable(Variable, "V", min=0.0)
         self.add_variable(Variable, "I")
+        self.add_variable(Variable, "V", min=0.0)
```
Comment (FJanssen-TNO) [inline src/mesido/pycml/component_library/milp/electricity/electricity_base.py:17]: add comment on the reason for this order of P, I and V

File: src/mesido/pycml/component_library/milp/gas/gas_base.py [hunk 3]
```diff
@@ -15,7 +15,7 @@ def __init__(self, name, **modifiers):
         # TODO: think of more elegant approach for Q_shadow, currently required to ensure that
         #  every port has a unique variable to make the correct port mapping
         self.add_variable(Variable, "Q")  # [m3/s]
-        self.add_variable(Variable, "Q_shadow")
+        # self.add_variable(Variable, "Q_shadow")
```
Comment (FJanssen-TNO) [inline src/mesido/pycml/component_library/milp/gas/gas_base.py:18]: Can we remove Q_shadow completely or do you want to keep in the comments?

File: src/mesido/pycml/component_library/milp/gas/gas_pipe.py [hunk 4]
```diff
@@ -45,7 +45,7 @@ def __init__(self, name, **modifiers):
             ((self.GasIn.mass_flow - self.GasOut.mass_flow) / (self.Q_nominal * self.density))
         )
         # # shadow Q for aliases
-        self.add_equation(((self.GasOut.Q_shadow - (self.GasIn.Q_shadow - 1.0e-3))))
+        # self.add_equation(((self.GasOut.Q_shadow - (self.GasIn.Q_shadow - 1.0e-3))))
```
Comment (FJanssen-TNO) [inline src/mesido/pycml/component_library/milp/gas/gas_pipe.py:48]: Q_shadow can be removed?

File: tests/test_logical_links.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                np.testing.assert_allclose(
+                    results[f"{node}.GasConn[{i_conn+1}].H"], results[f"{node}.H"], atol=1.0e-6
+                )
+            np.testing.assert_allclose(discharge_sum, 0.0, atol=1.0e-12)
+
+    def test_logical_links_network_hybrid(self):
+        """
+        This test checks if the logic implemented for logically linked energy system, Meaning an
+        energy system where assets are directly connected to each other without a network
+        (transport asset(s)) in between.
```
Comment (FJanssen-TNO) [inline tests/test_logical_links.py:113]: maybe state that hybrid refers to electricity and gas combined

File: tests/test_logical_links.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        np.testing.assert_allclose(
+            results["GasDemand_47d0.Gas_demand_mass_flow"]
+            + results["GasDemand_7978.Gas_demand_mass_flow"],
+            results["GasProducer_a977.Gas_source_mass_flow"]
+            + results["GasProducer_3573.Gas_source_mass_flow"],
+        )
+
+    def test_logical_links_nodes(self):
+
```
Comment (FJanssen-TNO) [inline tests/test_logical_links.py:77]: Add comment on test:
```suggestion
 This test checks if the logic implemented for a partially logically linked energy system. Meaning an
        energy system where assets are directly connected to each other without a network
        (transport asset(s)) in between. This specific test covers the logical connection between two nodes.
```

**Review by FJanssen-TNO — COMMENTED** [review]
Mainly comments on additional comments to be made in the code, besides that, it seems fine.

File: tests/test_logical_links.py
Comment (FJanssen-TNO) [inline tests/test_logical_links.py:1]: Do we need to add a test for heat?

File: src/mesido/esdl/asset_to_component_base.py [hunk 7]
```diff
@@ -355,28 +449,76 @@ def _set_electricity_current_nominal_and_max(
...
+        # elif asset.out_ports is None:
+        #     connected_port = asset.in_ports[0].connectedTo[0]
+        #     q_max = self._port_to_q_max.get(connected_port, None)
+        #     if q_max is not None:
+        #         self._set_q_max(asset, q_max)
+        #         return q_max
+        #     else:
+        #         raise _RetryLaterException(
+        #             f"Could not determine max discharge for {asset.asset_type} '{asset.name}'"
+        #         )
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/mesido/esdl/asset_to_component_base.py:518]: done

File: src/mesido/esdl/asset_to_component_base.py [hunk 8]
```diff
@@ -451,70 +625,95 @@ def _get_connected_q_nominal(self, asset: Asset) -> Union[float, Dict]:
...
+                        else max([asset.attributes.get(key, -1) for key in self.__power_keys])
+                        / (
+                            get_density(asset.name, port.carrier)
+                            * get_internal_energy(asset.name, port.carrier)
+                            / 1.0e3
+                        )
+                    )
+                    if q_nominal is not None:
+                        self._set_q_nominal(asset, q_nominal)
+                        return q_nominal
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/mesido/esdl/asset_to_component_base.py:678]: I think not. We want to have the port with a heat of gas commodity and skip the ports with electricity when retrieving a flow nominal. for example skip the electricity in port of a e-boiler

File: src/mesido/esdl/esdl_model_base.py (hunk 1 above)
Comment (jimrojerTNO) [reply src/mesido/esdl/esdl_model_base.py:501]: done

File: src/mesido/pycml/component_library/milp/electricity/electricity_base.py (hunk 2 above)
Comment (jimrojerTNO) [reply src/mesido/pycml/component_library/milp/electricity/electricity_base.py:17]: actually the order doesn't matter any more in the current implementation

File: src/mesido/pycml/component_library/milp/gas/gas_base.py (hunk 3 above)
Comment (jimrojerTNO) [reply src/mesido/pycml/component_library/milp/gas/gas_base.py:18]: should be removed now

File: tests/test_logical_links.py (hunk 6 above)
Comment (jimrojerTNO) [reply tests/test_logical_links.py:79]: done

File: tests/test_logical_links.py (hunk 5 above)
Comment (jimrojerTNO) [reply tests/test_logical_links.py:117]: done

File: src/mesido/esdl/esdl_heat_model.py [hunk 9]
```diff
@@ -1904,7 +1843,7 @@ def convert_gas_demand(self, asset: Asset) -> Tuple[Type[GasDemand], MODIFIERS]:
             # *hydrogen_specfic_energy),
             density=density,
             GasIn=dict(
-                Q=dict(min=0.0, max=q_max, nominal=q_nominal),
+                Q=dict(min=0.0, nominal=q_nominal),
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/mesido/esdl/esdl_heat_model.py:1846]: happened at some stage in debugging. However, we might want to keep it that way as we do not set q_max for the other type of demands.

File: tests/test_logical_links.py
Comment (jimrojerTNO) [reply tests/test_logical_links.py:1]: added

**Review by jimrojerTNO — APPROVED** [review]

**Pull request merged**